pkg/infrastructure/gcp/clusterapi/iam.go

```diff
@@ -102,48 +106,56 @@ func AddServiceAccountRoles(ctx context.Context, projectID, serviceAccountID str
 		return fmt.Errorf("failed to create resourcemanager service: %w", err)
 	}
 
-	policy, err := getPolicy(ctx, service, projectID)
-	if err != nil {
-		return err
+	backoff := wait.Backoff{
+		Duration: 2 * time.Second,
+		Jitter:   1.0,
+		Steps:    5,
 	}
-
-	member := fmt.Sprintf("serviceAccount:%s", serviceAccountID)
-	for _, role := range roles {
-		err = addMemberToRole(policy, role, member)
-		if err != nil {
-			return fmt.Errorf("failed to add role %s to %s: %w", role, member, err)
+	// Get and set the policy in a backoff loop and return errors if any.
+	return wait.ExponentialBackoffWithContext(ctx, backoff, func(ctx context.Context) (bool, error) {
```

Comment:
Instead of just returning whatever `ExponentialBackoff` returns, it might be a good idea to "save" the last error and return that instead of context errors like `context deadline exceeded` which doesn't give us much clue on what went wrong in the retries.

Reply:
You can use https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait#Interrupted to check whether the wait error was from the context.

Reply:
Good point, will fix.